rfcs/0000-plan-dynanism.md

```diff
@@ -0,0 +1,261 @@
+      \[There are probably other ways to make store paths that end in `.drv` that aren't valid derivations, so we could make the simpler change of lifting this restriction entirely without breaking invariants. But I'm fine keeping it for the wrong sorts of derivations as a useful guard rail.\]
+
+2. Extend the CLI to take advantage of such derivations:
+
+   We hopefully will soon allow CLI "installable" args in the form
+   ```
+   single-installable ::= <path> ! <output-name>
+   ```
+   where the first path is a derivation, and the second is the output we want to build.
+
+   We should generalize the grammar like so:
+   ```
+   single-installable ::= <single-installable> ! <output-name>
+                       |  <path>
+
+   multi-installable  ::= <single-installable>
+                       |  <single-installable> ! *
+   ```
+
+   Plain paths just mean that path itself is the goal, while `!` indexing indicates one more outputs of the derivation to the left of the `!` is the goal.
+
+   > For example,
+   > ```
+   > nix build /nix/store/…foo.drv
+   > ```
+   > would just obtain `/nix/store/…foo.drv` and not build it, while
+   > ```
+   > nix build /nix/store/…foo.drv!*
+   > ```
+   > would obtain (by building or substituting) all its outputs.
+   > ```
+   > nix build /nix/store/…foo.drv!out!out
+   > ```
+   > would obtain the `out` output of whatever derivation `/nix/store/…foo.drv!out` produces.
+
+   Now that we have `path` vs `path!*`, we also don't need `--derivation` as a disambiguator, and so that should be removed along with all the complexity that goes with it.
+   (`toDerivedPathsWithHints` in the the nix commands should always be pure functions and not consult the store.)
+
+3. Extend the scheduler and derivation dependencies similarly:
+
+  - Derivations can depend on the outputs of derivations that are themselves derivation outputs.
+    The scheduler will substitute derivations to simplify dependencies as computed derivations are built, just like how floating content addressed derivations are realized.
+
+  - Missing derivations get their own full fledged goals so they can be built, not just fetched from substituters.
+
+4. Add a new `outputOf` primop:
+
+   `builtins.outputOf drv outputName` produces a placeholder string with the appropriate string context to access the output of that name produced by that derivation.
+   The placeholder string is quite analogous to that used for floating content addressed derivation outputs.
+   \[With just floating content-addressed derivations but no computed derivations, derivations are always known statically but their outputs aren't.
+   With this RFC, since derivations themselves can floating CA derivation outputs, we also might not know them statically, so we need "deep" placeholders to account for arbitrary layers of dynamism.
+   This also corresponds to the use of arbitrary many `!` in the CLI.\]
```

Comment:
It also makes it so that you don't need `!` on the CLI if the expression accounts for the correct number of `outputOf`s, iiuc.

Reply:
How will a package definition look when it uses this functionality? We'll want to provide certain package attributes "statically", so they don't depend on anything in the derivation, so as not to necessitate IFD for `meta`, etc. I don't see an obvious way to extend `mkDerivation` to do this. Even if we do, its name is not quite appropriate anymore, because it mks more than one derivation.

Perhaps we could define "package" as an attrset with outputs, `meta`, etc, according to current conventions. We could then have a new `mkPackage` to take care of the fixpoint, overriding, meta and passthru, but leave derivation-specific aspects to the old `mkDerivation` (`buildInputs`, etc). It might look like:

```nix
# mkPackage: tie the knot, return the `public` attrset, add `overridePackage` function
mkPackage (self: {
  public = finalOutputs // {
    # other public attrs we care about
    version = "1.2.3";
    tests = f self;
    meta = {
      description = /* ... */;
    };
  };
  # An arbitrary internal attribute to allow overriding a lockfile
  lockFile = /* ... */;
  # Generate the derivation
  drvDrv = stdenv.mkDerivation {
    inherit (self.public) version;
    inherit (self) lockFile;
    outputHashMethodThingy = "text"; # output is a derivation (using wrong syntax, sorry)
    /* ... */
  };
  finalOutputs =
    # lib.packageOutputsOf wraps builtins.outputOf returning outputs in the usual format
    lib.packageOutputsOf ["out"] self.drvDrv;
})
```

`mkPackage` allows any number of derivations to be part of the package. It removes most of the need for PR https://github.com/NixOS/nixpkgs/pull/119942. Also `public` is more self-evident than `passthru` and we get overridable protected attributes. Fewer public attributes by default is a good thing. If you need to access internals, you can `overridePackage` to expose them if you really have to.

It probably needs some refinement, but I feel like this could be a good change for Nixpkgs.